src/node_locks.cc

```diff
@@ -0,0 +1,312 @@
+#include "node_locks.h"
+#include <algorithm>
+#include <memory>
+#include <utility>
+#include "env-inl.h"
+#include "node.h"
+#include "node_internals.h"
+#include "v8.h"
+
+namespace node {
+namespace worker {
+
+using v8::Array;
+using v8::Boolean;
+using v8::Context;
+using v8::Function;
+using v8::FunctionCallbackInfo;
+using v8::FunctionTemplate;
+using v8::HandleScope;
+using v8::Isolate;
+using v8::Local;
+using v8::Null;
+using v8::Number;
+using v8::Object;
+using v8::Promise;
+using v8::String;
+using v8::TryCatch;
+using v8::Undefined;
+using v8::Value;
+
+LockManager LockManager::current_;
+
+void LockManager::SnapshotTask::AfterThreadPoolWork(int status) {
+  HandleScope scope(env_->isolate());
+  LockManager* manager = LockManager::GetCurrent();
+
+  Mutex::ScopedLock lock(manager->work_mutex_);
+
+  manager->Snapshot(env_, promise_.Get(env_->isolate()));
+}
+
+void LockRequest::AfterThreadPoolWork(int status) {
+  HandleScope scope(env_->isolate());
+  LockManager* manager = LockManager::GetCurrent();
+
+  Mutex::ScopedLock lock(manager->work_mutex_);
+
+  // If steal is true, then run these steps:
+  if (steal_) {
+    // For each lock of held:
+    for (auto const& lock : manager->held_locks_) {
```

Comment:
@addaleax ... just thinking out loud here... would an `std::unordered_map<std::string,std::deque<std::unique_ptr<Lock>>` work better here to avoid having to loop through all locks and perform a string comparison on each?

Reply:
That’s very much possible – I haven’t really looked at the PR on that level, but it sounds like something that makes sense to me.